src/shogun/machine/gp/SparseVGInferenceMethod.cpp

```diff
@@ -0,0 +1,436 @@
+	Map<MatrixXd> eigen_inv_Lm(m_inv_Lm.matrix, m_inv_Lm.num_rows, m_inv_Lm.num_cols);
+	//invLm = Lm\eye(model.m); 
+	eigen_inv_Lm=Luu.matrixU().solve(MatrixXd::Identity(m_kuu.num_rows, m_kuu.num_cols));
+
+	m_Knm_inv_Lm=SGMatrix<float64_t>(m_ktru.num_cols, m_ktru.num_rows);
+	Map<MatrixXd> eigen_Knm_inv_Lm(m_Knm_inv_Lm.matrix, m_Knm_inv_Lm.num_rows, m_Knm_inv_Lm.num_cols);
+	//KnmInvLm = model.Knm*invLm;      
+	eigen_Knm_inv_Lm=(eigen_ktru.transpose()*CMath::exp(m_log_scale*2.0))*eigen_inv_Lm;
+
+	m_Tmm=SGMatrix<float64_t>(m_kuu.num_rows, m_kuu.num_cols);
+	Map<MatrixXd> eigen_C(m_Tmm.matrix, m_Tmm.num_rows, m_Tmm.num_cols);
+	//C = KnmInvLm'*KnmInvLm; 
+	eigen_C=eigen_Knm_inv_Lm.transpose()*eigen_Knm_inv_Lm;
+
+	m_inv_La=SGMatrix<float64_t>(m_kuu.num_rows, m_kuu.num_cols);
+	Map<MatrixXd> eigen_inv_La(m_inv_La.matrix, m_inv_La.num_rows, m_inv_La.num_cols);
+	//A = sigma2*eye(model.m) + C;    
+	LLT<MatrixXd> chol_A(m_sigma2*MatrixXd::Identity(m_kuu.num_rows, m_kuu.num_cols)+eigen_C);
+
+	//La = chol(A);
+	//invLa =  La\eye(model.m);  
+	eigen_inv_La=chol_A.matrixU().solve(MatrixXd::Identity(m_kuu.num_rows, m_kuu.num_cols));
+
+	//L=-invLm*invLm' +  sigma2*(invLm*invLa*invLa'*invLm');
+	m_L=SGMatrix<float64_t>(m_kuu.num_rows, m_kuu.num_cols);
+	Map<MatrixXd> eigen_L(m_L.matrix, m_L.num_rows, m_L.num_cols);
+	eigen_L=eigen_inv_Lm*(
+		m_sigma2*eigen_inv_La*eigen_inv_La.transpose()-MatrixXd::Identity(m_kuu.num_rows, m_kuu.num_cols)
+		)*eigen_inv_Lm.transpose();
+
+	//TrK = - (0.5/sigma2)*(model.TrKnn  - sum(diag(C)) );
+	m_trk=-0.5/(m_sigma2)*(eigen_ktrtr_diag.array().sum()*CMath::exp(m_log_scale*2.0)
+		-eigen_C.diagonal().array().sum());
+}
+
+void CSparseVGInferenceMethod::update_alpha()
+{
+	Map<MatrixXd> eigen_Knm_inv_Lm(m_Knm_inv_Lm.matrix, m_Knm_inv_Lm.num_rows, m_Knm_inv_Lm.num_cols);
+	Map<MatrixXd> eigen_inv_La(m_inv_La.matrix, m_inv_La.num_rows, m_inv_La.num_cols);
+	Map<MatrixXd> eigen_inv_Lm(m_inv_Lm.matrix, m_inv_Lm.num_rows, m_inv_Lm.num_cols);
+
+	SGVector<float64_t> y=((CRegressionLabels*) m_labels)->get_labels();
+	Map<VectorXd> eigen_y(y.vector, y.vlen);
+	SGVector<float64_t> m=m_mean->get_mean_vector(m_features);
+	Map<VectorXd> eigen_m(m.vector, m.vlen);
+
+	//yKnmInvLm = (model.y'*KnmInvLm);  
+	//yKnmInvLmInvLa = yKnmInvLm*invLa;     
+	VectorXd y_cor=eigen_y-eigen_m;
+	VectorXd eigen_y_Knm_inv_Lm_inv_La_transpose=eigen_inv_La.transpose()*(
+		eigen_Knm_inv_Lm.transpose()*y_cor);
+	//alpha = invLm*invLa*yKnmInvLmInvLa'; 
+	m_alpha=SGVector<float64_t>(m_kuu.num_rows);
+	Map<VectorXd> eigen_alpha(m_alpha.vector, m_alpha.vlen);
+	eigen_alpha=eigen_inv_Lm*eigen_inv_La*eigen_y_Knm_inv_Lm_inv_La_transpose;
+
+	m_yy=y_cor.dot(y_cor);
+	//F3 = (0.5/sigma2)*(yKnmInvLmInvLa*yKnmInvLmInvLa');
+	m_f3=0.5*eigen_y_Knm_inv_Lm_inv_La_transpose.dot(eigen_y_Knm_inv_Lm_inv_La_transpose)/m_sigma2;
+}
+
+void CSparseVGInferenceMethod::update_deriv()
+{
+	Map<MatrixXd> eigen_inv_La(m_inv_La.matrix, m_inv_La.num_rows, m_inv_La.num_cols);
```

Comment:
For all those complicated looking blocks, a one sentence comment on what happens on a high level might be good for future developers to understand this